### DISCUSSION TITLE
.chezmoidata and templating

### DISCUSSION THREAD

**Discussion by conor-sometimes** [twpayne/chezmoi#2892] [Q&A] [answered]
Hi all, Just recently started using chezmoi and so far it's perfect for my use cases.
I apologise if this was asked before but I was wondering if templating is able to be used in .chezmoidata

I'm trying to use chezmoi to set variables depending on hostnames as I'm using chezmoi to manage different machines (4 in the future when I dig into chezmoi on windows).

my current .chezmoidata looks like:
```
NAME = "my name"
EMAIL = "personal email"

[MACBOOK]
        EMAIL = "work email"
        GPG_KEY = "work gpg ID"
        SSH_KEY = "work ssh key location"

[SERVER]
        GPG_KEY = "server gpg ID"
        SSH_KEY = "server ssh key location"

[DESKTOP]
        GPG_KEY = "desktop gpg ID"
        SSH_KEY = "desktop ssh key location"
```

And then propagate the variables into the various configs/scripts that need them.

In those files i use the following templating logic to assign them based on my machines hostnames
**gpg.conf.tmpl**
```
{{ if eq .chezmoi.os "darwin" }}
default-key = {{ .MACBOOK.GPG_KEY }}
{{ else if eq .chezmoi.hostname "SERVER" }}
default-key = {{ .DESKTOP.GPG_KEY }}
{{ else if eq .chezmoi.hostname "DESKTOP" }}
default-key = {{ .SERVER.GPG_KEY }}
{{- end }}
```

Which is fine in all honesty, but with the hostname checking logic in .chezmoidata instead it could look like this:
**.chezmoidata**
```
{{ if eq .chezmoi.os "darwin" }}
EMAIL = "work email"
GPG_KEY = "work gpg ID"
SSH_KEY = "work ssh key location"
{{ else if eq .chezmoi.hostname "SERVER" }}
EMAIL = "personal email"
GPG_KEY = "server gpg ID"
SSH_KEY = "server ssh key location"
{{ else if eq .chezmoi.hostname "DESKTOP" }}
EMAIL = "personal email"
GPG_KEY = "desktop gpg ID"
SSH_KEY = "desktop ssh key location"
{{- end }}
```
And then we could just use the variable {{ .EMAIL }}, for example in configs and it would be correct based on hostname/os.

Would this be possible?
I'm guessing it's not as when I run chezmoi with the above setup I get the following error
`chezmoi: template: dot_config/git/config.tmpl:4:11: executing "dot_config/git/config.tmpl" at <.NAME>: map has no entry for key "NAME"`

Apologies if I'm abusing chezmoi to do something that it's not intended to do, thanks.

**Comment by nandalopes** [comment, marked as answer]
I'd use a different approach, combining `chezmoidata` and `.chezmoi.<format>.tmpl`.

```toml
# chezmoi data
[email]
MACBOOK = "work email"
DESKTOP = "desktop email"
SERVER = "server email"
```

```toml
# .chezmoi.toml.tmpl
[data]
computer = "DESKTOP"
```

Then when using at template, you can do:

`{{ dig .computer "" .email }}`

See [dig documentation](http://masterminds.github.io/sprig/dicts.html#dig).

**Reply by conor-sometimes** [reply]
That's a genius solution, many thanks!
